src/zarr/core/buffer/core.py

```diff
@@ -467,9 +466,12 @@ def all_equal(self, other: Any, equal_nan: bool = True) -> bool:
             # Handle None fill_value for Zarr V2
             return False
         # use array_equal to obtain equal_nan=True functionality
+        # Note from Ryan: doesn't this lead to a huge amount of unnecessary memory allocation on every single chunk?
+        # Since fill-value is a scalar, isn't there a faster path than allocating a new array for fill value
+        # every single time we have to write data?
         _data, other = np.broadcast_arrays(self._data, other)
         return np.array_equal(
-            self._data, other, equal_nan=equal_nan if self._data.dtype.kind not in "US" else False
+            self._data, other, equal_nan=equal_nan if self._data.dtype.kind not in "USTO" else False
```

Comment:
Is the `T` in the `dtype.kind`s here a typo? I don't see it at https://numpy.org/doc/2.0/reference/generated/numpy.dtype.kind.html.

Reply:
`np.dtypes.StringDType().kind` => `'T'`

Reply:
Actually, it's interesting that you flagged that here, because StringDType has some cool support for missing data: https://numpy.org/devdocs/user/basics.strings.html#missing-data-support

Not quite sure how we should be leveraging that in Zarr.